plugin/storage/es/spanstore/reader.go

```diff
@@ -393,15 +393,36 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []model.TraceID, st
 	return traces, nil
 }
 
+func traceIDQuery(traceID model.TraceID) elastic.Query {
+	traceIDStr := traceID.String()
+	// TODO remove in newer versions, added in Jaeger 1.16
+	// read ID without leading zeros
+	// https://github.com/jaegertracing/jaeger/pull/1956 added leading zeros to IDs
+	if traceIDStr[0] == '0' {
+		var legacyTraceID string
+		if traceID.High == 0 {
+			legacyTraceID = fmt.Sprintf("%x", traceID.Low)
+		} else {
+			legacyTraceID = fmt.Sprintf("%x%016x", traceID.High, traceID.Low)
+		}
+		return elastic.NewBoolQuery().
+			Should(elastic.NewTermQuery(traceIDField, traceIDStr).Boost(2), elastic.NewTermQuery(traceIDField, legacyTraceID))
+	}
+	return elastic.NewTermQuery(traceIDField, traceIDStr)
+}
+
 func convertTraceIDsStringsToModels(traceIDs []string) ([]model.TraceID, error) {
+	traceIDsMap := map[model.TraceID]bool{}
 	traceIDsModels := make([]model.TraceID, len(traceIDs))
 	for i, ID := range traceIDs {
 		traceID, err := model.TraceIDFromString(ID)
 		if err != nil {
 			return nil, errors.Wrap(err, fmt.Sprintf("Making traceID from string '%s' failed", ID))
 		}
-
-		traceIDsModels[i] = traceID
+		if _, ok := traceIDsMap[traceID]; !ok {
+			traceIDsMap[traceID] = true
+			traceIDsModels[i] = traceID
```

Comment:
This may leave gaps in the final array. We need to allocate it with 0 size and use `append()`

Reply:
would that cause issues? I wanted to avoid reallocation.

Reply:
it's still pre-allocated to the max possible length, so no reallocation